File: src/test/java/ru/romanow/jpa/web/dao/EntityDaoImpl.java
package ru.romanow.jpa.web.dao;

import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Repository;

import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.CriteriaQuery;
import jakarta.persistence.criteria.Root;
import java.util.List;
import java.util.Optional;

@Repository
public class EntityDaoImpl
        implements EntityDao {

    @PersistenceContext
    private EntityManager entityManager;

    @Override
    public <T> Optional<T> findById(@NotNull Integer id, Class<T> cls) {
        final CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
        final CriteriaQuery<T> query = criteriaBuilder.createQuery(cls);
        final Root<T> root = query.from(cls);
        query.select(root).where(criteriaBuilder.equal(root.get("id"), id));
        return entityManager
                .createQuery(query)
                .setMaxResults(1)
                .getResultList()
                .stream()
                .findFirst();
    }

    @Override
    public <T> List<T> findAll(Class<T> cls) {
        final CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
        final CriteriaQuery<T> query = criteriaBuilder.createQuery(cls);
        final Root<T> root = query.from(cls);
        return entityManager
                .createQuery(query.select(root))
                .getResultList();
    }
}

File: src/test/java/ru/romanow/jpa/dao/EntityDaoImpl.kt
package ru.romanow.jpa.dao

import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext
import org.springframework.stereotype.Repository
import java.util.*

@Repository
class EntityDaoImpl(@PersistenceContext private val entityManager: EntityManager) : EntityDao {

    override fun <T> findById(id: Int, cls: Class<T>): Optional<T> {
        val criteriaBuilder = entityManager.criteriaBuilder
        val query = criteriaBuilder.createQuery(cls)
        val root = query.from(cls)
        query.select(root).where(criteriaBuilder.equal(root.get<Any>("id"), id))
        return entityManager
            .createQuery(query)
            .setMaxResults(1)
            .resultList
            .stream()
            .findFirst()
    }

    override fun <T> findAll(cls: Class<T>): List<T> {
        val criteriaBuilder = entityManager.criteriaBuilder
        val query = criteriaBuilder.createQuery(cls)
        val root = query.from(cls)
        return entityManager
            .createQuery(query.select(root))
            .resultList
    }
}
